### PR TITLE
[BugFix][Spec Decode] Preserve DeepSeek V4 DSpark heads during weight loading

### LINKED ISSUE #17094: [Bug]: DeepSeek V4 DSpark target weights overwrite draft head/HC depending on checkpoint load order
### Your current environment

Reproduced on two ARM64 hosts, each with 8 x Ascend 910B3 (A2), TP=8, DSpark K=7, draft eager, target FULL_DECODE_ONLY. Both EP off and EP on are affected.

Sanitized environment summary collected using `collect_env.py` (deployment-specific addresses and environment variables omitted):

```text
OS: Ubuntu 22.04.5 LTS (aarch64)
Host kernel: Linux 5.10.0-60.18.0.50.oe2203.aarch64
Python: 3.12.13
PyTorch: 2.10.0+cpu
torch_npu: 2.10.0.post4
Transformers: 5.14.1
CANN: 9.1.0
vLLM: 0.28.0, commit 2cf0a6915ce544dc493a0990f2ea38d81601128a
vLLM Ascend: commit 44f9bf53d93fe6a5b1362b7d81a050d3b0cd997d
Image: quay.io/ascend/vllm-ascend:nightly-main
Pinned local ARM64 image ID: sha256:d3e8b2b20036638a285872871696dcded219f6ee751e8070412f57f4189b187d
Model: Eco-Tech/DeepSeek-V4-Flash-0731-w8a8
ModelScope revision: 5012d242bcbfb2686702c3fe0c1bfea83cc7d3fb (2026-09-07)
```

The real-model run used `--quantization ascend` and `--model-loader-extra-config '{"enable_multithread_load":true,"num_threads":128}'`. The minimal loader reproducer below also fails on current Ascend main `0003e1b75`; multithreading is **not** required to reproduce the ordering bug.

### 🐛 Describe the bug

`DSparkDeepseekV4ForCausalLM.load_weights()` maps the target checkpoint's top-level head/embed/HC keys and the `mtp.*` draft keys to the same destination parameters. When a target key arrives after its MTP counterpart, it silently replaces the trained draft parameter. The result depends on checkpoint iteration/shard completion order.

For example, both `mtp.2.head.weight` and `head.weight` map to `lm_head.weight`. Loading the former and then the latter leaves the draft model using the **target** output head. The same problem affects `embed.weight` and `hc_head_fn`, `hc_head_base`, `hc_head_scale`. The verifier still rejects incorrect proposals, so requests can succeed while speculative decoding loses most of its benefit.

#### Minimal reproducer

Run this in a vLLM Ascend Python environment. It calls the actual loader with tiny CPU tensors and mocked parallel-state accessors; it does not download a checkpoint, initialize a model, or run inference:

```python
from types import MethodType, SimpleNamespace
from unittest.mock import patch

import torch
from vllm_ascend.models.deepseek_v4 import dspark

for draft_first in (True, False):
    head = torch.nn.Parameter(torch.zeros(2, 2), requires_grad=False)
    model = SimpleNamespace(
        rotation_path=None,
        config=SimpleNamespace(num_attention_heads=8, num_hidden_layers=43),
        model=SimpleNamespace(num_dspark_layers=3, get_expert_mapping=lambda: []),
        named_parameters=lambda: [("lm_head.weight", head)],
    )
    model._remap_dspark_name = MethodType(
        dspark.DSparkDeepseekV4ForCausalLM._remap_dspark_name, model
    )
    weights = [
        ("mtp.2.head.weight", torch.full((2, 2), 7.0)),
        ("head.weight", torch.ones(2, 2)),
    ]
    if not draft_first:
        weights.reverse()
    with (
        patch.object(dspark, "get_tensor_model_parallel_world_size", return_value=1),
        patch.object(dspark, "get_tensor_model_parallel_rank", return_value=0),
    ):
        dspark.DSparkDeepseekV4ForCausalLM.load_weights(model, iter(weights))
    print(f"{draft_first=}: loaded={head[0, 0].item()}, expected=7.0")
```

Observed on unpatched main:

```text
draft_first=True: loaded=1.0, expected=7.0
draft_first=False: loaded=7.0, expected=7.0
```

Expected: `7.0` in both cases. Explicit MTP weights should take precedence independently of iteration order. Global keys should remain fallbacks when the corresponding MTP key is absent.

#### Why the September checkpoint exposes the problem

We compared two local snapshots of the same ModelScope model:

| Checkpoint | Shards | Quantization description | Loader behavior |
|---|---:|---|---|
| Earlier snapshot | 74 | Declares `optional.quarot.rotation_map.global_rotation` | `rotation_path` is set; target head/embed are skipped |
| Revision `5012d242` | 75 | `optional: {}` | `rotation_path=None`; target head/embed may overwrite draft parameters |

Both snapshots contain `optional/quarot.safetensors`, but `get_rotation_path()` uses the metadata rather than file existence. Neither snapshot has `is_rot_used=true`. Full tensor SHA256 comparisons show that the draft LM head, draft embedding and all three draft HC tensors are identical between these snapshots. Within each snapshot, target and draft head/embed/HC tensors differ. Thus overwriting a draft tensor with its target counterpart is not an equivalent substitution.

The quantization scheme also changed, so copying the earlier rotation metadata into the new snapshot is not a justified fix. HC loading is not guarded by `rotation_path` and can still be overwritten on the rotation-enabled path.

#### Regression history

For quantized checkpoints without a rotation path, the direct regression is #13506, merged on 2026-08-07, commit `683b6c009d57714eaf9363f369c7ed4fa0084f5c`. It changed the target head/embed guard from `not self.has_own_*` to `not self.rotation_path` and added target HC loading. Executing the actual loader bodies from this commit and its parent confirms the before/after behavior. The earlier #11431 implementation already had an analogous head/embed ordering issue on the non-quantized path.

#### Real-model evidence and proposed fix

- Before repair, all 8 TP draft-head shards matched the target `head.weight`, rather than `mtp.2.head.weight`; one audited rank also had overwritten HC parameters.
- Holding the running model and test input fixed, repairing draft head/HC restored mean acceptance length from approximately 1.1 to 5.1. This was a diagnostic intervention, followed by a persistent loader fix and fresh model reloads.
- After the persistent fix, all 8 draft-head shards and all draft HC parameters matched the checkpoint. No runtime repair was used for the performance retest.
- C1 pure decode TPOT recovered from **48.70 to 12.20 ms (EP off)** and **49.16 to 12.50 ms (EP on)** on the same checkpoint. These are request-mean TPOT values excluding TTFT/prefill. The private replay request set was held fixed; the repaired result uses two consecutive warm 64-request batches per concurrency/configuration. Across repaired C1/C8 runs, 512 formal requests completed without errors.

Proposed fix: remember which destination head/embed/HC parameters were loaded from `mtp.*`, ignore later target fallbacks for those parameters, and still allow later MTP tensors to overwrite earlier fallbacks. This needs only a small set of parameter names, not checkpoint buffering, and does not change sampling or acceptance rules.

Related symptom report: #13640; its root cause is not established, so this issue does not claim they are identical. #15824 fixes a different Qwen3 DSpark rotation-metadata path and does not address this loader overwrite.

AI assistance: diagnosis, code and regression tests were prepared with OpenAI Codex assistance during the deployment investigation.

### PR DESCRIPTION
**Pull request by X0Ken**
### What this PR does / why we need it?

Fixes #17094.

DeepSeek V4 DSpark maps both target and MTP-specific head/embed/HC checkpoint keys to the same parameters. If `mtp.2.head.weight` arrives before the target `head.weight`, the latter silently overwrites the trained draft head. Unordered shard completion can therefore collapse acceptance while requests still succeed.

This change gives MTP-specific parameters priority independently of loading order:

- Track the five head/embed/HC destination names already loaded from `mtp.*`.
- Skip subsequent target fallbacks for those parameters; allow later MTP weights to replace earlier fallbacks.
- Preserve standalone and partial-MTP fallback behavior and the existing rotation guards, without buffering checkpoint tensors.

The September 7 W8A8 checkpoint's `optional: {}` exposes the target head/embed fallback path. This is a loader correctness fix, not a change to checkpoint metadata, the forward path, or acceptance rules. The issue contains a tiny checkpoint-free reproducer, the checkpoint comparison, and the regression history (#13506).

### Does this PR introduce _any_ user-facing change?

Yes: checkpoints containing dedicated MTP heads consistently use those heads regardless of shard order, restoring speculative acceptance. There are no new flags or API changes. Target weights remain fallbacks when the corresponding MTP weights are absent.

### How was this patch tested?

- Added `tests/ut/model_executor/test_dspark_head_loading.py`: **262 passed**, covering explicit target-first/draft-first order, 128 seeded shuffles, rotation present/absent, target-only fallback, and partial MTP heads. The PR's complete model module was imported into the existing Ascend test environment; only parallel-state accessors are mocked by the tests.
- Negative control using the unmodified main model module: **243 failed, 19 passed**. The minimal issue reproducer produces `1.0 / 7.0` before the fix and `7.0 / 7.0` after it.
- `bash format.sh ci`: passed (rerun after the shellcheck bootstrap encountered a concurrent download/extraction error).
- Hardware validation of this same loader fix on the pinned nightly stack: two 8 x 910B3 hosts, TP8 / K7, EP off/on; fresh model reloads verified all 8 TP draft-head shards and all HC parameters against the checkpoint, without runtime weight repair.
- Real-workload C1/C8: **512 formal requests passed** across two consecutive warm 64-request batches per configuration/concurrency. C1 request-mean **pure decode TPOT**, excluding TTFT/prefill, recovered from **48.70 to 12.20 ms (EP off)** and **49.16 to 12.50 ms (EP on)**. This is recovery from corrupted draft weights, not a speedup over an already-correct model.

Hardware measurements used Ascend `44f9bf53d` / vLLM `2cf0a6915` with this loader fix. This PR is based on main `0003e1b75`; its loader/remapper method ASTs are identical to the hardware-validated fix. The complete latest-main serving stack has not been re-benchmarked.

To run the regression suite in a compatible development environment:

```bash
python -m pytest -q tests/ut/model_executor/test_dspark_head_loading.py
```

AI assistance: OpenAI Codex assisted with diagnosis, implementation, tests, and this description. The commit includes AI attribution and DCO sign-off.

- vLLM main: https://github.com/vllm-project/vllm/commit/84030bbe3d74d99bad477a3d2e37a973ccd8865c

### CODE REVIEW

**Comment by X0Ken** [comment]
Hi @drslark and @realliujiaxu, could you please review this DSpark weight-loading fix? Your work on #11431 and #13506 is directly relevant to the loader behavior addressed here.

CPU unit tests, pre-commit, and DCO have passed. The precision/full test jobs have not run because the PR is waiting for a test-enabling label. Could a maintainer please add `ready-precise` to trigger the selected tests? The regression tests and pinned-stack hardware validation are documented in the PR description.

Thank you!
